File: code/Java/Chapter8/8-3/SelectionSort.java
import java.util.Arrays;

public class SelectionSort {
    public static void main(String[] args) {
        int[] arr = {5, 8, 6, 3, 9, 2, 1, 7};
        selectionSort(arr);
        System.out.println(Arrays.toString(arr));
    }

    /**
     * Selection Sort (original)
     */
    public static int[] selectionSort(int[] arr) {
        for (int i = 0; i < arr.length - 1; i++) {
            int minIndex = i;
            for (int j = i + 1; j < arr.length; j++) {
                if (arr[j] < arr[minIndex]) {
                    minIndex = j;
                }
            }
            int temp = arr[i];
            arr[i] = arr[minIndex];
            arr[minIndex] = temp;
        }

        return arr;
    }

    /**
     * Selection Sort (optimized)
     */
    // public static int[] selectionSort(int[] arr) {
    //     for (int i = 0; i < arr.length / 2; i++) {
    //         int minIndex = i;
    //         int maxIndex = arr.length - i - 1;

    //         for (int j = i; j < arr.length - i; j++) {
    //             if (arr[j] < arr[minIndex]) {
    //                 minIndex = j;
    //             }
    //             if (arr[j] > arr[maxIndex]) {
    //                 maxIndex = j;
    //             }
    //         }

    //         int temp = arr[i];
    //         arr[i] = arr[minIndex];
    //         arr[minIndex] = temp;

    //         // In case of i == maxIndex before swap(arr[i], arr[minIndex]), it's now at minIndex.
    //         if (i == maxIndex) {
    //             maxIndex = minIndex;
    //         }

    //         temp = arr[arr.length - i - 1];
    //         arr[arr.length - i - 1] = arr[maxIndex];
    //         arr[maxIndex] = temp;
    //     }

    //     return arr;
    // }
}

File: code/Python/Chapter8/8-3/selection_sort.py
def selection_sort(lst):
    """Selection Sort (original v1.0)"""
    for i in range(len(lst)):
        min_index = i
        for j in range(i + 1, len(lst)):
            if lst[j] < lst[min_index]:
                min_index = j
        lst[i], lst[min_index] = lst[min_index], lst[i]
        
    return lst

# def selection_sort(lst):
#     """Selection Sort (optimized v2.0)"""
#     for i in range(len(lst) // 2):
#         min_index = i
#         max_index = len(lst) - i - 1

#         for j in range(i, len(lst) - i):
#             if lst[j] < lst[min_index]:
#                 min_index = j
#             if lst[j] > lst[max_index]:
#                 max_index = j
            
#         lst[i], lst[min_index] = lst[min_index], lst[i]

#         # In case of i == max_index before swap(lst[i], lst[min_index]), it's now at min_index.
#         if i == max_index:
#             max_index = min_index
        
#         lst[len(lst) - i - 1], lst[max_index] = lst[max_index], lst[len(lst) - i - 1]

#     return lst

lst = [5, 8, 6, 3, 9, 2, 1, 7]
print(selection_sort(lst))